### PR TITLE
Fix crashes when multiple Vulkan layers are loaded

### PR DESCRIPTION
**Pull request by florian-kuebler**
They were two conception problems when in vkCreateDevice
that may lead to a crash when another Vulkan layer gets
loaded.

1. Calling vkGetInstanceProcAddr with VK_NULL_HANDLE
In order to call vkCreateDevice down the chain, we
retrieve the function pointer using the vkGetInstanceProcAddr
function for vkCreateDevice.
As we did not care about the instance, we pass in "VK_NULL_HANDLE",
as described in the Vulkan layer tutorial and almost all other
Vulkan layer. However, vkCreateDevice is not a global function,
but an instance function, so per API, this call will return
a nullptr and the subsequent call to the function pointer will
fail. While most ICDs and layers will return the right function
pointer anyways, Renderdoc returns null in that case.

So we do the right thing, and call vkGetInstanceProcAddr with
the actual instance.

2. `u.pLayerInfo->pfnNextGetInstanceProcAddr` is not always
   returning the function pointer down the chain.
For most instance functions, like "vkEnumerateDeviceExtensionProperties"
the Vulkan loader assumes that after the "vkCreateInstance" call
you are not doing another lookup using the pfnNextGetInstanceProcAddr
pointer, but instead create a dispatch table.
If layer A is before our Vulkan layer, we might end up in
getting the function pointer of layer A and fail.

So we use the cached dispatch table instead.
Some more context can be found in http://b/243079253.

Test: load the Vulkan layer together with RenderDoc on a couple
      of games.

Bug: http://b/243079253

### CODE REVIEW

File: src/OrbitVulkanLayer/VulkanLayerController.h [hunk 1]
```diff
@@ -704,15 +704,14 @@ class VulkanLayerController {
...
-  void AddRequiredDeviceExtensionNameIfMissing(
-      const VkDeviceCreateInfo* create_info, VkPhysicalDevice physical_device,
-      PFN_vkGetInstanceProcAddr next_get_instance_proc_addr_function, const char* extension_name,
-      std::vector<std::string>* output) {
+  void AddRequiredDeviceExtensionNameIfMissing(const VkDeviceCreateInfo* create_info,
+                                               VkPhysicalDevice physical_device,
+                                               const char* extension_name,
+                                               std::vector<std::string>* output) {
     auto raw_enumerate_device_extension_properties_function =
         // Pass a valid instance, as following the spec. we are not allowed to use nullptr here.
```
Comment (dpallotti) [inline src/OrbitVulkanLayer/VulkanLayerController.h:712]: Is this still valid?

File: src/OrbitVulkanLayer/VulkanLayerController.h [hunk 2]
```diff
@@ -704,15 +704,14 @@ class VulkanLayerController {
...
         absl::bit_cast<PFN_vkEnumerateDeviceExtensionProperties>(
-            next_get_instance_proc_addr_function(dispatch_table_.GetInstance(physical_device),
```
Comment (dpallotti) [inline src/OrbitVulkanLayer/VulkanLayerController.h:714]: > For most instance functions, like "vkEnumerateDeviceExtensionProperties"
the Vulkan loader assumes that after the "vkCreateInstance" call
you are not doing another lookup using the pfnNextGetInstanceProcAddr
pointer, but instead create a dispatch table.
If layer A is before our Vulkan layer, we might end up in
getting the function pointer of layer A and fail.

I'm trying to understand this better. So after calling `vkCreateInstance` the loader might return something unexpected from `pfnNextGetInstanceProcAddr`? Why? What does "might" mean in this case? Is it clear why a different function pointer is returned, but only if a layer is loaded before our layer?

File: src/OrbitVulkanLayer/VulkanLayerController.h [hunk 3]
```diff
@@ -181,8 +180,9 @@ class VulkanLayerController {
 
     // Need to call vkCreateInstance down the chain to actually create the
     // instance, as we need it to be alive in the create instance dispatch table.
-    auto create_device_function = absl::bit_cast<PFN_vkCreateDevice>(
-        next_get_instance_proc_addr_function(VK_NULL_HANDLE, "vkCreateDevice"));
+    auto create_device_function =
+        absl::bit_cast<PFN_vkCreateDevice>(next_get_instance_proc_addr_function(
+            dispatch_table_.GetInstance(physical_device), "vkCreateDevice"));
```
Comment (dpallotti) [inline src/OrbitVulkanLayer/VulkanLayerController.h:185]: > per API, this call will return
a nullptr and the subsequent call to the function pointer will
fail. While most ICDs and layers will return the right function
pointer anyways, Renderdoc returns null in that case.

Maybe more of a curiosity, but: Does this mean that when passing `VK_NULL_HANDLE` then nullptr should be returned, while the right function should be returned when passing an instance? Or does it mean that nullptr _can_ be returned when passing `VK_NULL_HANDLE` (basically undefined behavior in this case)?

**Review by dpallotti — COMMENTED** [review]
Forgive if they are silly questions, I forgot most about Vulkan so I want to try and understand better. Code looks completely reasonable.

**Comment by hysw** [comment]
> 1. ... so per API, this call will return a nullptr

Technically the returned thing is `undefined`, so usually implementation returns a valid pointer. But sometime it returns nullptr. 

> 2. Vulkan loader assumes that after the "vkCreateInstance" call you are not doing another lookup using the pfnNextGetInstanceProcAddr

You can only get the proper address for `vkGetInstanceProcAddr`, `vk_layerGetPhysicalDeviceProcAddr`, `vkCreateInstance`, `vkCreateDevice` and some debug util functions

File: src/OrbitVulkanLayer/VulkanLayerController.h (hunk 2 above)
Comment (hysw) [reply src/OrbitVulkanLayer/VulkanLayerController.h:714]: See [loader_init_instance_core_dispatch_table](https://github.com/KhronosGroup/Vulkan-Loader/blob/16d5d8f25452f207d32d3dec95c0531b8527728e/loader/loader.c#L4724) at the end of loader trampoline, it fills the dispatch table with the top layer function pointer, replacing the icd ones (if the layer defined those function)

File: src/OrbitVulkanLayer/VulkanLayerController.h (hunk 3 above)
Comment (hysw) [reply src/OrbitVulkanLayer/VulkanLayerController.h:185]: It's undefined.
https://registry.khronos.org/vulkan/specs/1.3-extensions/man/html/vkGetInstanceProcAddr.html#_description

File: src/OrbitVulkanLayer/VulkanLayerController.h (hunk 2 above)
Comment (dpallotti) [reply src/OrbitVulkanLayer/VulkanLayerController.h:714]: Thanks!

File: src/OrbitVulkanLayer/VulkanLayerController.h (hunk 3 above)
Comment (dpallotti) [reply src/OrbitVulkanLayer/VulkanLayerController.h:185]: And thanks!

File: src/OrbitVulkanLayer/VulkanLayerController.h (hunk 1 above)
Comment (dpallotti) [reply src/OrbitVulkanLayer/VulkanLayerController.h:712]: I guess this is the only open comment for me.

**Review by dpallotti — APPROVED** [review]

File: src/OrbitVulkanLayer/VulkanLayerController.h (hunk 1 above)
Comment (florian-kuebler) [reply src/OrbitVulkanLayer/VulkanLayerController.h:712]: It is still valid, but, I would argue that it now confuses more than it adds, so I am going to remove it. It is pretty clear what is going on here, anyways.

**Pull request merged**